pkg/auth/providers/ldap/ldap_client_test.go

```diff
@@ -0,0 +1,323 @@
+						ServiceAccountPassword:          DummySAUPassword,
+						UserObjectClass:                 UserObjectClassName,
+					},
+				},
+				caPool: &x509.CertPool{},
+			},
+			want: v3.Principal{
+				ObjectMeta: v1.ObjectMeta{
+					Name: "providername_user://ldap.test.domain",
+				},
+				PrincipalType: "user",
+				Me:            true,
+			},
+			want1: []v3.Principal{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "providername_group://ldap.test.domain",
+					},
+					PrincipalType: "user",
+					Me:            true,
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "user login with invalid credentials",
+			fields: fields{
+				userMGR: mockUserManager{
+					hasAccess: false,
+				},
+				tokenMGR:   &tokens.Manager{},
+				caPool:     &x509.CertPool{},
+				userScope:  "providername_user",
+				groupScope: "providername_group",
+			},
+			args: args{
+				lConn: &mockLdapConn{
+					canAuthenticate: false,
+				},
+				credential: &v32.BasicLogin{
+					Username: DummyUsername,
+					Password: DummyPassword,
+				},
+				config: &v3.LdapConfig{
+					LdapFields: v32.LdapFields{
+						ServiceAccountDistinguishedName: DummySAUsername,
+						ServiceAccountPassword:          DummySAUPassword,
+						UserObjectClass:                 UserObjectClassName,
+					},
+				},
+				caPool: &x509.CertPool{},
+			},
+			want:    v3.Principal{},
+			want1:   nil,
+			wantErr: true,
+		},
+		{
+			name: "user login without access permissions",
+			fields: fields{
+				userMGR: mockUserManager{
+					hasAccess: false,
+				},
+				tokenMGR:   &tokens.Manager{},
+				caPool:     &x509.CertPool{},
+				userScope:  "providername_user",
+				groupScope: "providername_group",
+			},
+			args: args{
+				lConn: newMockLdapConnClient(),
+				credential: &v32.BasicLogin{
+					Username: DummyUsername,
+					Password: DummyPassword,
+				},
+				config: &v3.LdapConfig{
+					LdapFields: v32.LdapFields{
+						ServiceAccountDistinguishedName: DummySAUsername,
+						ServiceAccountPassword:          DummySAUPassword,
+						UserObjectClass:                 UserObjectClassName,
+					},
+				},
+				caPool: &x509.CertPool{},
+			},
+			want:    v3.Principal{},
+			want1:   nil,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &ldapProvider{
+				ctx:                   tt.fields.ctx,
+				authConfigs:           tt.fields.authConfigs,
+				secrets:               tt.fields.secrets,
+				userMGR:               tt.fields.userMGR,
+				tokenMGR:              tt.fields.tokenMGR,
+				certs:                 tt.fields.certs,
+				caPool:                tt.fields.caPool,
+				providerName:          tt.fields.providerName,
+				testAndApplyInputType: tt.fields.testAndApplyInputType,
+				userScope:             tt.fields.userScope,
+				groupScope:            tt.fields.groupScope,
+			}
+			got, got1, err := p.loginUser(tt.args.lConn, tt.args.credential, tt.args.config, tt.args.caPool)
```

Comment:
nit: the loginUser call sites calls `got` -> **userPrincipal**, and `got1` -> **groupPrincipals**. I would use those names here, and use **wantUser** and **wantGroups** in the test table, just for clarity